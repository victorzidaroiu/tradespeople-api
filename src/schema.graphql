type Query {
  findJobsInRange(tradespersonId: Int): FindJobsResponse
}

type Mutation {
  addTradesperson(input: AddTradespersonInput): AddTradespersonResponse
  addJob(input: AddJobInput): AddJobResponse
}

type FindJobsResponse {
  success: Boolean,
  jobs: [Job]
}

type Job {
  description: String!,
  latitude: Float,
  longitude: Float
}

input AddTradespersonInput {
  name: String!, 
  latitude: Float!,
  longitude: Float!
}

input AddJobInput {
  description: String!, 
  latitude: Float!,
  longitude: Float!,
  maxTradespersonDistance: Float!
}

type AddTradespersonResponse {
  success: Boolean,
  tradespersonId: Int
}

type AddJobResponse {
  success: Boolean,
  jobId: Int
}
