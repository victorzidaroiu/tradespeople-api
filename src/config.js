export default {
  db: {
    storage: './db/main.db',
  },
  port: 3000,
};
