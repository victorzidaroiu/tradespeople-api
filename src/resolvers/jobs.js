import getDB from '../db';

export const addJob = async (_,
  {
    input: {
      description, longitude, latitude, maxTradespersonDistance,
    },
  }) => {
  const db = getDB();

  const insertId = await db.insertJob(description, longitude, latitude, maxTradespersonDistance);

  if (!insertId) {
    return {
      success: false,
    };
  }

  return {
    success: true,
    jobId: insertId,
  };
};

export const findJobsInRange = async (_, { tradespersonId }) => {
  const db = getDB();

  const jobs = await db.findJobsInRange(tradespersonId);

  if (!jobs) {
    return {
      success: false,
    };
  }

  return {
    success: true,
    jobs,
  };
};
