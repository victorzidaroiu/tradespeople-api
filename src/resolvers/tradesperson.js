import getDB from '../db';

// eslint-disable-next-line import/prefer-default-export
export const addTradesperson = async (_, { input: { name, longitude, latitude } }) => {
  const db = getDB();

  const insertId = await db.insertTradesperson(name, longitude, latitude);

  if (!insertId) {
    return {
      success: false,
    };
  }

  return {
    success: true,
    tradespersonId: insertId,
  };
};
