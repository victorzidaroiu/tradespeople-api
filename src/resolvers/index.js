import { addTradesperson } from './tradesperson';
import { addJob, findJobsInRange } from './jobs';

export default {
  Query: {
    findJobsInRange,
  },
  Mutation: {
    addTradesperson,
    addJob,
  },
};
