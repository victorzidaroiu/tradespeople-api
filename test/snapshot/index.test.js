/* eslint-disable import/no-extraneous-dependencies */
import 'babel-polyfill'
import { join } from 'path';
import { graphql } from 'graphql';
import { loadSchemaSync } from '@graphql-tools/load';
import { addResolversToSchema } from '@graphql-tools/schema';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import resolvers from '../../src/resolvers';
import getDb from '../../src/db';
import config from '../../src/config';

const schema = loadSchemaSync(join(__dirname, '../../src/schema.graphql'), {
  loaders: [
    new GraphQLFileLoader(),
  ],
});

const schemaWithResolvers = addResolversToSchema({
  schema,
  resolvers,
});

describe('Snapshots tests', () => {
  beforeEach(async () => {
    config.db.storage = ':memory:';
    const db = getDb();
    await db.connect();
    await db.createTables();
  });

  afterEach(async () => {
    config.db.storage = ':memory:';
    const db = getDb();
    await db.removeTables();
  });

  it('Create tradesperson', async () => {
    const response = await graphql(schemaWithResolvers, `
      mutation {
        addTradesperson(input:{ name: "Joe", longitude: 100, latitude: 100 }) {
          success
          tradespersonId
        }
      }
    `);

    expect(JSON.stringify(response)).toMatchSnapshot();
  });

  it('Create job', async () => {
    const response = await graphql(schemaWithResolvers, `
      mutation {
        addJob(input:{  description: "Job A", longitude: 101, latitude: 101, maxTradespersonDistance: 100 }) {
          success,
          jobId
        }
      }
    `);

    expect(JSON.stringify(response)).toMatchSnapshot();
  });

  it('Find tradesperson jobs', async () => {
    await graphql(schemaWithResolvers, `
      mutation {
        addTradesperson(input:{ name: "Joe", longitude: 100, latitude: 100 }) {
          success
          tradespersonId
        }
      }
    `);

    await graphql(schemaWithResolvers, `
      mutation {
        addJob(input:{  description: "Job X", longitude: 101, latitude: 101, maxTradespersonDistance: 100 }) {
          success,
          jobId
        }
      }
    `);

    await graphql(schemaWithResolvers, `
      mutation {
        addJob(input:{  description: "Job Y", longitude: 50, latitude: 50, maxTradespersonDistance: 75 }) {
          success,
          jobId
        }
      }
    `);

    const response = await graphql(schemaWithResolvers, `
      query {
        findJobsInRange(tradespersonId: 1) {
          success,
          jobs {
            description,
            latitude,
            longitude
          }
        }
      }
    `);

    expect(JSON.stringify(response)).toMatchSnapshot();
  });
});
