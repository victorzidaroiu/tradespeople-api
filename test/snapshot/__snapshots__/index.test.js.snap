// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots tests Create job 1`] = `"{\\"data\\":{\\"addJob\\":{\\"success\\":true,\\"jobId\\":1}}}"`;

exports[`Snapshots tests Create tradesperson 1`] = `"{\\"data\\":{\\"addTradesperson\\":{\\"success\\":true,\\"tradespersonId\\":1}}}"`;

exports[`Snapshots tests Find tradesperson jobs 1`] = `"{\\"data\\":{\\"findJobsInRange\\":{\\"success\\":true,\\"jobs\\":[{\\"description\\":\\"Job X\\",\\"latitude\\":101,\\"longitude\\":101},{\\"description\\":\\"Job Y\\",\\"latitude\\":50,\\"longitude\\":50}]}}}"`;
